### ISSUE TITLE
"Truncate Silence" fails

### ISSUE THREAD

**Issue by SteveDaulton** [audacity/audacity#3988]
### Bug description

"Truncate Silence" fails to remove silences.

### Steps to reproduce

1. Generate 30 second "Chirp".
2. Make 5 splits are approximately regular intervals
3. Move the clips so that there is a couple of seconds gap between the clips.
4. Select All
5. Apply "Truncate Silence" with these settings:
![Screenshot_2022-11-27_19-30-56](https://user-images.githubusercontent.com/11681097/204155817-5d78e397-24dd-40b8-b60e-83212c1a4b15.png)

### Expected behavior

The gaps to be reduced to 0.5 seconds.

### Actual behavior

The selection gets smaller, but the clips don't move.

### Audacity Version

current master / alpha / nightly build

### Operating system

Linux

### Additional context

Appears to be a regression against 3.2.1

**Comment by SteveDaulton** [comment]
Also happens in 3.2.1 with default Preferences.

**Comment by DavidBailes** [comment]
Looks like the behaviour of truncate silence has always depended on the setting preferences > track behaviors > editing a clip can move other clips. Since this setting has been disabled by default, then with default settings, the behaviour of truncate silence is as described in this issue.

**Comment by petersampsonaudacity** [comment]
It also leads to this rather odd effect of changing the selection as though the pseudo-silence gap had been removed:

### Before Truncate Silence with default settings
![image](https://user-images.githubusercontent.com/50205984/204260449-b0df967a-500b-44b7-9296-f9b27b9ccc19.png)

### After Truncate Silence with default settings
![image](https://user-images.githubusercontent.com/50205984/204260668-9c791663-a20f-4001-a647-4034ff734550.png)

**Comment by petersampsonaudacity** [comment]
Testing this also begs another _(to me somewhat odd)_ design decision.

Just why is the default "Truncate to" set to be 0.5 seconds?  Surely most users who want silence truncated want it truncated to zero?
![image](https://user-images.githubusercontent.com/50205984/204261931-124b151a-88fd-475e-a88c-c5e8d77a81e3.png)

**Comment by SteveDaulton** [comment]
> Just why is the default "Truncate to" set to be 0.5 seconds?

I believe that was chosen for reducing pauses in speech recordings. (Very handy for listening to lecture notes when there may be long pauses).

**Comment by petersampsonaudacity** [comment]
> Looks like the behaviour of truncate silence has always depended on the setting preferences > track behaviors > editing a clip can move other clips. Since this setting has been disabled by default, then with default settings, the behaviour of truncate silence is as described in this issue.

@DavidBailes  I will later add a note to the Truncate Silence page of the Manual warning the reader about this.

**Comment by LWinterberg** [comment]
How about this: 

1. Truncate Silence is moved to Edit > Remove Special
2. A new option is added to Edit > Audio Clips called "Remove Gaps" which removes gaps between clips (#3924)
3. Truncate Silence is given permission to move clips around again if clips live within its selection

**Comment by SteveDaulton** [comment]
> 1. Truncate Silence is moved to Edit > Remove Special

That may be easier for people to find (now that it is no longer visible by default in the Effect menu), but technically "Truncate Silence" is an "Effect" (it processes the selected audio) rather than an "Edit". While this may seem a bit pedantic, there are some important differences in how effects behave compared to edits, for example when applied to a 16-bit track, an effect will resample the audio to 32-bit float and back to 16-bit (with dither by default), whereas edits do not modify the underlying data.






> 2\. A new option is added to Edit > Audio Clips called "Remove Gaps" which removes gaps between clips (

I'm not a fan of duplicating functionality unless it is adding a new, unique benefit. As was (imo) previously the case with so many ways to zoom in and out, duplicate functionality increases clutter.




> 3\. Truncate Silence is given permission to move clips around again if clips live within its selection

I think that's the correct solution.

**Comment by LWinterberg** [comment]
> I'm not a fan of duplicating functionality unless it is adding a new, unique benefit.

The unique benefit would be that you could truncate silences between gaps while leaving the clips themselves intact with any silences that may be in them.

**Comment by SteveDaulton** [comment]
> The unique benefit would be that you could truncate silences between gaps while leaving the clips themselves intact with any silences that may be in them.

Considering that can be done more easily than ever now that clips have drag bars (drag handles), what's the common use case that makes it worth developing and adding as a built-in rather than an optional plug-in? (It could be done with a Nyquist Macro).

**Comment by petersampsonaudacity** [comment]
> The unique benefit would be that you could truncate silences between gaps while leaving the clips themselves intact with any silences that may be in them.

That already happens with Edit > Truncate Silence

### Before
![image](https://user-images.githubusercontent.com/50205984/204329581-32c87fc1-7bd7-4584-8cc1-f39621e81e89.png)

### After
![image](https://user-images.githubusercontent.com/50205984/204329879-09ecfac3-ddc9-45dc-b73d-660ddf9fb471.png)

**Comment by LWinterberg** [comment]
> That already happens with Edit > Truncate Silence

No. Truncate Silence does not leave the clips intact if there's silence within them. 
![image](https://user-images.githubusercontent.com/87814144/204334709-260e5c30-3122-4027-8e4f-349f6e81819e.png)
![image](https://user-images.githubusercontent.com/87814144/204334869-266fec9a-8cd5-4326-8f79-b03ce0ff0fb4.png)

> Considering that can be done more easily than ever now that clips have drag bars (drag handles), what's the common use case that makes it worth developing and adding as a built-in rather than an optional plug-in? (It could be done with a Nyquist Macro).

Anything longform has this problem

**Comment by SteveDaulton** [comment]
> Truncate Silence does not leave the clips intact if there's silence within them.

That's right, but it looks like quite a fringe case. Even the example you gave above can be handled by Truncate Silence (by setting the "Detect Silence > Duration" greater than the flat line silences).




> It could be done with a Nyquist Macro

I've had a look at this, and yes it can be done, though handling all possible edge cases gets a bit messy. I'm not convinced that there's sufficient demand to justify the development effort even for a Nyquist plug-in (developing in Nyquist is much quicker than C++)

**Comment by petersampsonaudacity** [comment]
> No. Truncate Silence does not leave the clips intact if there's silence within them.

Nor should it, as the silences in you example are part of the two clips.

**Comment by petersampsonaudacity** [comment]
> @DavidBailes I will later add a note to the Truncate Silence page of the Manual warning the reader about this.

I have added an advice div to the Truncate Silence page in the 3.2.2 alpha Manual for this., I added a P2 for the removal of that ednote for if and when this issue gets fixed:

![image](https://user-images.githubusercontent.com/50205984/204548580-84d3c553-3c48-45fd-a189-4a6540a36a6f.png)

**Comment by petersampsonaudacity** [comment]
Note that even if '''Editing a clip can move other clips''' is in it's default "'''off'''" setting then subsequent clips do not honor the off setting and get moved.

### Before
![image](https://user-images.githubusercontent.com/50205984/204551084-bffbbf1d-56eb-48e2-a04b-f270f3eaaa0a.png)

### After
![image](https://user-images.githubusercontent.com/50205984/204551386-78d721d5-e2f7-4cb7-b6af-bcb47d2bd0ff.png)


### Note carefully that using Delete to remove the silence _**does**_ honor the "off" setting
![image](https://user-images.githubusercontent.com/50205984/204551982-daeff4e7-b028-4663-a196-1f0f0f85772c.png)

![image](https://user-images.githubusercontent.com/50205984/204552205-e97c0f39-e0a0-45d6-bb2e-8613f47f7768.png)

**Comment by LWinterberg** [comment]
> That may be easier for people to find (now that it is no longer visible by default in the Effect menu), but technically "Truncate Silence" is an "Effect" (it processes the selected audio) rather than an "Edit". While this may seem a bit pedantic, there are some important differences in how effects behave compared to edits, for example when applied to a 16-bit track, an effect will resample the audio to 32-bit float and back to 16-bit (with dither by default), whereas edits do not modify the underlying data.

This is a "tail wagging the dog" approach. Menu structures should be made to meet the user expectations, not because of technicalities. Whether or not something might resample the track before execution certainly is not the guiding principle in finding a tool, and even if a user cares about resampling at all, surely the solution is to make truncate silence to not resample the track. Given that both trimming, silencing, splitting and joining are available from the edit menu, it's definitely possible.

> So maybe Truncate Silence should have an option added to its dialog for: Treat inter-Clip gaps as silence

I mocked up adding such an option to the truncate silence dialog, but it turns an already confusing dialog even more confusing: It would require a tri-state toggle (truncate silence only, truncate gaps only, truncate silence and gaps), where the dB threshold wouldn't have any effect on the gaps. The dialog already is quite cluttered for what it does, and if you do just want to move all clips in the project back-to-back, it creates massive cognitive load for what could be a dialog-free one-click operation.

**Comment by SteveDaulton** [comment]
> Menu structures should be made to meet the user expectations

Expectations may differ according to knowledge and experience, but I do see your point. "Silence Audio" in the Edit menu is technically an effect (generator), but everyone (including myself) seems to be happy with that. I'm just suggesting caution about implementing things in ways that are not technically correct.



> surely the solution is to make truncate silence to not resample the track

Processing in Audacity is always done with 32-bit float data. Changing that would  likely be a huge refactoring task with hardly any tangible benefits.

It could be  done by rewriting Truncate Silence as an actual "edit" operation, but then it would frequently create clicks at the edit points.


Personally I think the best solution would be:

1. Add a new Edit menu command to remove gaps between selected clips ("Remove Gaps"?).
(It may be a good idea to decide how / if Audacity will handle stereo clips that have different start / end times before developing this new command).
2. In Truncate Silence, remove the -80dB limit for the threshold level, or at least allow it to be set much lower.

(The reason that it has a  -80 dB limit is historic - The effect was originally developed by Lynn Allan in 2007 with the dB level set by a multi-choice menu. The downside of the multi-choice menu was that the number of options has to be limited, but the advantage was that the final option was "Off-Skip" (equivalent to -inf dB). Ideally it would be a true logarithmic slider with a range -inf to 0dB, but Audacity does not currently have a widget like that.)

**Comment by petersampsonaudacity** [comment]
Another really odd behavior is that if you turn on the **Editing a clip can move other clips** and then turn it off again the the "off" setting gets ignored.

1. clear Audacity settings folder
2. generate  a thirty second chirp
3. split it at 15 seconds
4. move the second clip to the right
5. Edit > Preferences >Tracks Behaviors 
6. Observe:  **Editing a clip can move other clips** is "off" - default setting
7. select all
8. Truncate Silence
9. Observe: nothing happens
10. Edit > Preferences >Tracks Behaviors 
11. Turn **Editing a clip can move other clips**  "on"
12. select all
13. Truncate Silence
14. Observe: the inter-clip gap is removed  
15. Undo
16. Edit > Preferences >Tracks Behaviors 
17. Turn **Editing a clip can move other clips**  "**off**"
18. select all 
19. Truncate Silence
20. Observe: the inter-clip gap is removed  

But at step 20 I would expect the same behavior as at Steps 8-9 - as we now have the "off" setting so clips should not move.

**Comment by DavidBailes** [comment]
> Another really odd behavior is that if you turn on the **Editing a clip can move other clips** and then turn it off again the the "off" setting gets ignored.

I ran into this really confusing bug a short time ago, and I've logged it: https://github.com/audacity/audacity/issues/3998

**Comment by DavidBailes** [comment]
> I ran into this really confusing bug a short time ago, and I've logged it: #3998

And as I've commented on that issue, other settings have the same bug.
